File: Labs/Lab 3/CS/TriangleApp/Triangle.cs
using System;

namespace TriangleLibrary
{
    public class Triangle
    {
        private double x1, y1, x2, y2, x3, y3;

        public Triangle()
        {
            x1 = y1 = 0;
            x2 = 1; y2 = 0;
            x3 = 0; y3 = 1;
        }

        public Triangle(double x1, double y1, double x2, double y2, double x3, double y3)
        {
            this.x1 = x1; this.y1 = y1;
            this.x2 = x2; this.y2 = y2;
            this.x3 = x3; this.y3 = y3;
        }

        public Triangle(Triangle other)
        {
            x1 = other.x1; y1 = other.y1;
            x2 = other.x2; y2 = other.y2;
            x3 = other.x3; y3 = other.y3;
        }

        public double Perimeter()
        {
            double a = Distance(x1, y1, x2, y2);
            double b = Distance(x2, y2, x3, y3);
            double c = Distance(x3, y3, x1, y1);
            return a + b + c;
        }
        public double Area()
        {
            return Math.Abs((x1 * (y2 - y3) + x2 * (y3 - y1) + x3 * (y1 - y2)) / 2.0);
        }

        public static Triangle operator +(Triangle t1, Triangle t2)
        {
            return new Triangle(
                t1.x1 + t2.x1, t1.y1 + t2.y1,
                t1.x2 + t2.x2, t1.y2 + t2.y2,
                t1.x3 + t2.x3, t1.y3 + t2.y3
            );
        }

        public static Triangle operator *(Triangle t, double scalar)
        {
            return new Triangle(
                t.x1 * scalar, t.y1 * scalar,
                t.x2 * scalar, t.y2 * scalar,
                t.x3 * scalar, t.y3 * scalar
            );
        }

        public override string ToString()
        {
            return $"P={Perimeter():F2}, A={Area():F2}";
        }

        private double Distance(double x1, double y1, double x2, double y2)
        {
            return Math.Sqrt(Math.Pow(x2 - x1, 2) + Math.Pow(y2 - y1, 2));
        }
    }
}

File: Labs/Lab 3/TS/Triangle.ts
export class Triangle {
  private x1: number;
  private y1: number;
  private x2: number;
  private y2: number;
  private x3: number;
  private y3: number;

  constructor();
  constructor(
    x1: number,
    y1: number,
    x2: number,
    y2: number,
    x3: number,
    y3: number
  );
  constructor(triangle: Triangle);
  constructor(
    x1?: number | Triangle,
    y1?: number,
    x2?: number,
    y2?: number,
    x3?: number,
    y3?: number
  ) {
    if (x1 instanceof Triangle) {
      this.x1 = x1.x1;
      this.y1 = x1.y1;
      this.x2 = x1.x2;
      this.y2 = x1.y2;
      this.x3 = x1.x3;
      this.y3 = x1.y3;
    } else if (
      x1 !== undefined &&
      y1 !== undefined &&
      x2 !== undefined &&
      y2 !== undefined &&
      x3 !== undefined &&
      y3 !== undefined
    ) {
      this.x1 = x1;
      this.y1 = y1;
      this.x2 = x2;
      this.y2 = y2;
      this.x3 = x3;
      this.y3 = y3;
    } else {
      this.x1 = this.y1 = 0;
      this.x2 = 1;
      this.y2 = 0;
      this.x3 = 0;
      this.y3 = 1;
    }
  }


  perimeter(): number {
    return (
      this.distance(this.x1, this.y1, this.x2, this.y2) +
      this.distance(this.x2, this.y2, this.x3, this.y3) +
      this.distance(this.x3, this.y3, this.x1, this.y1)
    );
  }

  area(): number {
    return Math.abs(
      (this.x1 * (this.y2 - this.y3) +
        this.x2 * (this.y3 - this.y1) +
        this.x3 * (this.y1 - this.y2)) /
        2
    );
  }


  scale(factor: number): Triangle {
    return new Triangle(
      this.x1 * factor,
      this.y1 * factor,
      this.x2 * factor,
      this.y2 * factor,
      this.x3 * factor,
      this.y3 * factor
    );
  }

  add(other: Triangle): Triangle {
    return new Triangle(
      this.x1 + other.x1,
      this.y1 + other.y1,
      this.x2 + other.x2,
      this.y2 + other.y2,
      this.x3 + other.x3,
      this.y3 + other.y3
    );
  }

  toString(): string {
    return `Triangle: P=${this.perimeter().toFixed(2)}, A=${this.area().toFixed(
      2
    )}`;
  }

  private distance(x1: number, y1: number, x2: number, y2: number): number {
    return Math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2);
  }
}
